### PR TITLE
Fix FileNotFoundError in diff_templateoverride view and add regressio…

### LINKED ISSUE #5404: [Feature]: Add regression tests for diff_templateoverride view
### Problem Statement

The project currently lacks regression tests for the diff_templateoverride view and related utilities. This makes it harder to detect regressions in behavior such as permission handling, diff rendering, and template validation.

### Proposed Solution

Add regression tests covering:

- Admin access to diff_templateoverride view
- Non-admin permission handling (403)
- Invalid template ID handling (404)
- Diff rendering validation

Also include tests for:
- PropertyDict merge and flatten behavior
- FixIEMiddleware response handling

This improves test coverage and ensures stability of these components.

### Alternatives Considered

Manual testing was considered but is not reliable for preventing regressions. Automated tests provide consistent validation.

### Additional Context

Related work already implemented in PR #4333.

### PR DESCRIPTION
**Pull request by sankalp771**
## Description

Fixes an unhandled `FileNotFoundError` in the `diff_templateoverride` view which previously caused a 500 Internal Server Error when the original template file was missing.

The fix ensures the view now correctly raises an `Http404`, aligning with expected Django behavior.

Additionally, comprehensive regression tests have been added to validate correct behavior across multiple scenarios and prevent future breakage.

---

## Related Issue

Closes #5404 

---

## Type of Change

* [x] Bug fix
* [ ] New feature
* [ ] Breaking change
* [ ] Documentation update
* [ ] Refactor / cleanup
* [ ] CI/CD or build change

---

## Testing

* [x] Existing tests pass
* [x] New tests added (7 total cases for DiffTemplateOverrideViewTest)
* [x] Manually verified

### Test Coverage Added

* Admin access returns 200
* Non-admin access returns 403
* Invalid template ID returns 404
* Diff output renders expected content
* Identical content handled safely
* Existing template file handled correctly
* Missing original template file returns 404 (**regression test**)

All tests passing:

```
7 passed in ~140s
```
image

---

## AI Disclosure

AI tools (Amazon Q and ChatGPT) were used to assist in identifying the issue, debugging environment setup, and drafting test cases. Final implementation, validation, and testing were performed manually.

---

## Checklist

* [x] Self-reviewed the code
* [ ] Updated documentation (not required)
* [x] No new warnings or errors introduced

### CODE REVIEW

**Comment by Srinidhi188** [comment]
Thanks @sankalp771! Glad to collaborate on this.

I had worked on a similar approach in my earlier PR (#4333), so I’ll review this PR and see if I can help improve tests or fix any remaining issues. Let me know if there’s anything specific I can contribute here!

**Comment by sankalp771** [comment]
Thanks @Srinidhi188! Appreciate you taking the time to review this.

Since you’ve worked on a similar approach before, your feedback would be really valuable. Please feel free to suggest any improvements or edge cases I might have missed — happy to iterate on it.

**Comment by Srinidhi188** [comment]
> Thanks @Srinidhi188! Appreciate you taking the time to review this.
> 
> Since you’ve worked on a similar approach before, your feedback would be really valuable. Please feel free to suggest any improvements or edge cases I might have missed — happy to iterate on it.

Nice work on handling the FileNotFoundError and adding comprehensive tests!

One small suggestion: for the identical content case, should we also assert that the diff output is minimal or empty (instead of just checking status code)? This could help ensure the diff rendering behaves correctly when there are no changes.

Also, do you think it would be useful to add a test for empty template content to cover another edge case?

**Comment by Srinidhi188** [comment]
Thanks for the updates!

I noticed the PR currently has merge conflicts with the main branch. Would you like me to help resolve them or test after rebasing?

Also, once CI runs, I can help verify if all edge cases (like identical content and missing files) behave as expected.

**Comment by Srinidhi188** [comment]
Thanks for the updates!

Regarding the Copilot suggestions:

- For the test file creation, using a temporary directory (e.g., TemporaryDirectory or override_settings) sounds like a safer approach to avoid interfering with existing templates.

- For the email validator, I agree that escaping the domain properly (using re.escape) would make the regex more accurate.

- Also, the DirectorEmailValidator and migration changes seem slightly outside the main scope of this PR (which focuses on diff_templateoverride). It might be worth separating them into a different PR to keep this one focused.

Happy to help test once conflicts are resolved and CI runs.

File: esp/esp/utils/tests.py [hunk 1]
```diff
@@ -523,6 +528,108 @@ def test_fast_path_no_data_colon(self):
...
+        self.user = ESPUser.objects.create_user(username="user", password="pass")
+
+        # Create template directory + file
+        self.template_dir = os.path.join(settings.PROJECT_ROOT, "templates")
+        os.makedirs(self.template_dir, exist_ok=True)
+
+        self.template_path = os.path.join(self.template_dir, "test_template.html")
+
+        with open(self.template_path, "w") as f:
+            f.write("Original Content")
```
[the remark this answers is not in the document]
Comment (sankalp771) [reply esp/esp/utils/tests.py]: The fix is clear. I'll use tempfile.TemporaryDirectory + override_settings so the test never touches the real templates/ directory.

[the remark this answers is not in the document]
Comment (sankalp771) [reply esp/esp/utils/tests.py]: tempfile.TemporaryDirectory() — creates a fully isolated dir that auto-cleans up, unique per test run, never collides with anything

override_settings(PROJECT_ROOT=self._tmpdir.name) — makes the view read from the temp dir, not the real templates/

Template name changed from test_template.html to test_diff_template.html — unique name, zero chance of clobbering a real template

tearDown calls self._override.disable() then self._tmpdir.cleanup() — guaranteed full cleanup in correct order

No more os.makedirs(..., exist_ok=True) on the real project directory

**Comment by sankalp771** [comment]
image
test passed @willgearty  !

**Review by sankalp771 — COMMENTED** [review]
This mostly solves all concerns of the attached issue

**Comment by kkbrum** [comment]
@sankalp771 can you fix the failing test and merge conflicts?

**Comment by sankalp771** [comment]
Hi @kkbrum ,
will do it soon

**Review by Oval17 — APPROVED** [review]
As the fix was already on main, only tests we can add here.
Fixed lints and conflicts
LGTM

**Comment by sankalp771** [comment]
@willgearty @Oval17 
Thanks for attending this PR!! missed contributing in gsoc to this repo as I didn't got selected but loved contributing !
